File: coding/src/com/algos/core/li15_dp/revised/Subsequence_01Knapsack.java
package  com.algos.core.li15_dp.revised;

public class Subsequence_01Knapsack {

	public static void main(String[] args) {
		Subsequence_01Knapsack obj = new Subsequence_01Knapsack();
		
		int[] weights = { 1, 2, 3, 4, 5 };
		int[] values = { 5, 3, 5, 3, 2 };
		int n = 4, wt = 10;
		System.out.println(obj.knapsack(weights, values, n, wt));
		System.out.println(obj.knapsack(weights, values, n, wt));	// n, W
	}
	
	//  Problem: https://bit.ly/3KHpP3v
	//  
	//  n\w	0 	1	2	3	4	5	6	7	8	9	10
	//  1(5)	0	5	5	5	5	5	5	5	5	 5	5
	//  2(3)	0	5	5	8	8	8	8	8	8	8	8	
	//  3(5)	0	5	5	8	8	8	13	13	13	13	13
	//  4(3)	0	5	5	8	8	8	13	13	13	13	16
	//  5(2)	0	5	5	5	8	8	13	13	13	13	16
	//
	// Time: O(n*Wmax)
    // Space: O(n*Wmax)
	public int knapsack(int[] weight, int[] value, int n, int wmax) {
		int[][] val = new int[n][wmax+1];
		for(int ind=0; ind<n; ind++) {
			for(int w=0; w<=wmax; w++) {	
				if(ind == 0) {
					if(weight[0] <= w)			// for every wt w, we can pick 0th element
						val[0][w] = value[0];
				}else {
					int notPick = 0 + val[ind-1][w];
					int pick = Integer.MIN_VALUE;
					if(weight[ind] <= w)
						pick = value[ind]+val[ind-1][w-weight[ind]];
					val[ind][w] = Math.max(notPick, pick);
				}
			}
		}
		return val[n-1][wmax];
	}

	// n and w defines the state
	// O(N*Wmax) where n=number of elements and w is max weight
	public int knapsack2(int[] weight, int[] value, int ind, int wmax) {
		if (ind == 0){
			if(weight[0] <= wmax)
				return value[0];
			return 0;
		}
			
		int notPick = 0 + knapsack(weight, value, ind - 1, wmax);
		int pick = Integer.MIN_VALUE;
		if(weight[ind] <= wmax)
			pick = value[ind] + knapsack(weight, value, ind - 1, wmax - weight[ind]);

		return Math.max(notPick, pick);
	}
}

File: dsa/patterns/15_dp/python/subsequence_01_knapsack.py
# DP / Knapsack
#
# Problem: https://bit.ly/3KHpP3v
# Idea:
#   - State: (n, W) = max value using first n items with max weight W
#   - Pick or not pick each item
# Time: O(n*Wmax)
# Space: O(n*Wmax)

def knapsack_recursive(weights, values, ind, wmax):
    if ind == 0:
        if weights[0] <= wmax:
            return values[0]
        return 0

    notPick = knapsack_recursive(weights, values, ind - 1, wmax)
    pick = float('-inf')
    if weights[ind] <= wmax:
        pick = values[ind] + knapsack_recursive(weights, values, ind - 1, wmax - weights[ind])

    return max(notPick, pick)

def knapsack_dp(weights, values, wmax):
    n = len(weights)
    dp = [[0] * (wmax + 1) for _ in range(n)]

    for ind in range(n):
        for w in range(wmax + 1):
            if ind == 0:
                if weights[0] <= w:
                    dp[0][w] = values[0]
            else:
                notPick = dp[ind - 1][w]
                pick = float('-inf')
                if weights[ind] <= w:
                    pick = values[ind] + dp[ind - 1][w - weights[ind]]
                dp[ind][w] = max(notPick, pick)

    return dp[n - 1][wmax]


# -------- Test --------
weights = [1, 2, 3, 4, 5]
values = [5, 3, 5, 3, 2]
n = 4
wmax = 10

print(knapsack_recursive(weights, values, n - 1, wmax))  # Recursive
print(knapsack_dp(weights, values, wmax))                # DP
